bluefog/common/mpi_controller.cc

```diff
@@ -555,6 +626,33 @@ void MPIController::NeighborAllreduce(std::vector<TensorTableEntry>& entries) {
   timeline_ptr->ActivityEndAll(entries);
   const void* fused_input_data = buffer_data;
 
+  const void* weighted_fused_input_data = nullptr;
+  if (first_entry.dst_weighting_enabled) {
+    // Generate weighted data fusion for sending
+    timeline_ptr->ActivityStartAll(entries, "MEMCPY_IN_WEIGHT_FUSION_BUFFER");
+    void* weight_buffer_data;
+    MemcpyInWeightFusionBuffer(weight_buffer_data, first_entry.send_neighbors->size(),
+                               fused_input_data, num_elements, element_size,
+                               first_entry.context, first_entry.device);
+    int64_t offset = 0;
+    for (size_t i = 0; i < first_entry.send_neighbors->size(); ++i) {
+      double dst_weight = first_entry.send_weights->at(i);
+      void* weight_buffer_data_offset = (uint8_t*)weight_buffer_data + offset;
+      if (first_entry.device == CPU_DEVICE_ID) {
+        ScaleCPUBuffer(dst_weight, weight_buffer_data_offset, num_elements,
+                       first_entry.tensor->dtype());
+      } else {
+#if HAVE_CUDA
+        ScaleBufferCudaImpl(dst_weight, weight_buffer_data_offset, num_elements,
+                            first_entry.tensor->dtype(), mpi_ctx_.stream);
+#endif
+      }
+      offset += num_elements * element_size;
+    }
+    weighted_fused_input_data = weight_buffer_data;
+    timeline_ptr->ActivityEndAll(entries);
+  }
```

Comment:
Maybe it good to extract this out to a utility function.
`
const void* weighted_fused_input_data = MaybeGenerateWeightedFusedInputData(...);
`

Reply:
Done.